cpp/src/io/parquet/reader_impl_preprocess_utils.cu

```diff
@@ -0,0 +1,535 @@
+                       device_span<PageInfo> pages,
+                       rmm::cuda_stream_view stream)
+{
+  auto const num_pages = pages.size();
+  auto page_indexes    = cudf::detail::make_host_vector<page_index_info>(num_pages, stream);
+
+  for (size_t c = 0, page_count = 0; c < chunks.size(); c++) {
+    auto const& chunk = chunks[c];
+    CUDF_EXPECTS(chunk.h_chunk_info != nullptr, "Expected non-null column info struct");
+    auto const& chunk_info = *chunk.h_chunk_info;
+    size_t start_row       = 0;
+    page_count += chunk.num_dict_pages;
+    for (size_t p = 0; p < chunk_info.pages.size(); p++, page_count++) {
+      auto& page      = page_indexes[page_count];
+      page.num_rows   = chunk_info.pages[p].num_rows;
+      page.chunk_row  = start_row;
+      page.num_nulls  = chunk_info.pages[p].num_nulls.value_or(0);
+      page.num_valids = chunk_info.pages[p].num_valid.value_or(0);
+      page.str_bytes  = chunk_info.pages[p].var_bytes_size.value_or(0);
+
+      start_row += page.num_rows;
+    }
+  }
+
+  auto d_page_indexes = cudf::detail::make_device_uvector_async(
+    page_indexes, stream, cudf::get_current_device_resource_ref());
+
+  auto iter = thrust::make_counting_iterator<size_type>(0);
+  thrust::for_each(
+    rmm::exec_policy_nosync(stream), iter, iter + num_pages, copy_page_info{d_page_indexes, pages});
+}
+
+std::string encoding_to_string(Encoding encoding)
+{
+  switch (encoding) {
+    case Encoding::PLAIN: return "PLAIN";
+    case Encoding::GROUP_VAR_INT: return "GROUP_VAR_INT";
+    case Encoding::PLAIN_DICTIONARY: return "PLAIN_DICTIONARY";
+    case Encoding::RLE: return "RLE";
+    case Encoding::BIT_PACKED: return "BIT_PACKED";
+    case Encoding::DELTA_BINARY_PACKED: return "DELTA_BINARY_PACKED";
+    case Encoding::DELTA_LENGTH_BYTE_ARRAY: return "DELTA_LENGTH_BYTE_ARRAY";
+    case Encoding::DELTA_BYTE_ARRAY: return "DELTA_BYTE_ARRAY";
+    case Encoding::RLE_DICTIONARY: return "RLE_DICTIONARY";
+    case Encoding::BYTE_STREAM_SPLIT: return "BYTE_STREAM_SPLIT";
+    case Encoding::NUM_ENCODINGS:
+    default: return "UNKNOWN(" + std::to_string(static_cast<int>(encoding)) + ")";
+  }
+}
+
+[[nodiscard]] std::string encoding_bitmask_to_str(uint32_t encoding_bitmask)
+{
+  std::bitset<32> bits(encoding_bitmask);
+  std::string result;
+
+  for (size_t i = 0; i < bits.size(); ++i) {
+    if (bits.test(i)) {
+      auto const current = static_cast<Encoding>(i);
+      if (!is_supported_encoding(current)) { result.append(encoding_to_string(current) + " "); }
+    }
+  }
+  return result;
+}
+
+[[nodiscard]] std::string list_unsupported_encodings(device_span<PageInfo const> pages,
+                                                     rmm::cuda_stream_view stream)
+{
+  auto const to_mask = cuda::proclaim_return_type<uint32_t>([] __device__(auto const& page) {
+    return is_supported_encoding(page.encoding) ? uint32_t{0} : encoding_to_mask(page.encoding);
```

Comment:
This is the only change I made: `0U -> uint32_t{0}`. Same change to the second last argument to `thrust::transform_reduce` just below